File: src/api/integrations/auditboard.py
"""AuditBoard GRC client — file a scanner finding as a GRC issue.

Differs from the Jira integration in one way that drives the whole design:
AuditBoard has no prefill-a-form URL. Creating an issue is an authenticated
API write, so a credential has to live somewhere. It lives here, server-side,
read from settings and never returned to the browser.

Vocabulary was read from the live instance rather than assumed. Two things
that reference confirms, and that are easy to get wrong:

*   ``issue_rating_id`` is **not** a severity. The values are control-testing
    outcomes — Cleared, Tested - Remains Deficient, New Deficiency, N/A.
    Mapping a code severity onto it produces nonsense.
*   ``deficiency_level_id`` is the severity-shaped field: Recommendation - No
    Deficiency, Control Deficiency, Significant Deficiency, Material Weakness.

"Significant Deficiency" and "Material Weakness" are SOX terms of art. This
module will not assert either from a scanner severity on its own — see
``DEFICIENCY_LEVEL_BY_SEVERITY`` — because a static-analysis hit is evidence
of a control weakness, not a determination of one. The caller may override.
"""

import logging
from typing import Any, Dict, List, Optional

import requests

from ..config import settings

logger = logging.getLogger(__name__)

API_PATH = "/api/v1/issues"

# ---------------------------------------------------------------------------
# Reference vocabularies, read from the live instance.
# Verified against GET /api/v1/deficiency_levels and /api/v1/issue_ratings.
# ---------------------------------------------------------------------------

DEFICIENCY_LEVELS: Dict[int, str] = {
    1: "Recommendation - No Deficiency",
    2: "Control Deficiency",
    3: "Significant Deficiency",
    4: "Material Weakness",
    5: "TBD",
    6: "N/A",
}

# Control-testing outcomes. Recorded so nobody maps severity onto them again.
ISSUE_RATINGS: Dict[int, str] = {
    1: "Cleared",
    2: "Tested - Insufficient Sample",
    3: "Tested - Remains Deficient",
    4: "Not Tested - Management Remediated",
    5: "Remediation In Progress",
    6: "New Deficiency",
    7: "TBD",
    8: "N/A",
    9: "Closed",
}

# Our severity ramp onto deficiency levels.
#
# Deliberately conservative at the top: critical and high both land on
# "Control Deficiency". Escalation to Significant Deficiency or Material
# Weakness is a judgement a person makes with business context, not something
# a Semgrep rule earns automatically. Pass ``deficiency_level_id`` explicitly
# to go higher.
DEFICIENCY_LEVEL_BY_SEVERITY: Dict[str, int] = {
    "critical": 2,
    "high": 2,
    "medium": 2,
    "low": 1,
    "info": 1,
}

# How the issue was identified. 7 is "Other", the honest answer for a scanner:
# this was not found by Internal Audit, External Audit or Management.
# Verified against GET /api/v1/issue_idents.
ISSUE_IDENTS: Dict[int, str] = {
    4: "Internal Audit",
    5: "External Audit",
    6: "Management",
    7: "Other",
    8: "TBD",
    9: "N/A",
    10: "IT - Infrastructure",
}
DEFAULT_ISSUE_IDENT_ID = 7

# Statuses this instance actually uses, counted over the whole issues
# collection (28,644 records, paged through) rather than a first page:
#
#   Closed 22,940 · Inactive 2,232 · Pending Remediation 1,643 · Open 1,488
#   Remediated 248 · Canceled 93
#
# An earlier version of this set held only Open, Closed and Remediated and said
# those were the only values across 918 issues. Both figures were wrong: the
# page[number] parameter was being ignored, so the same first page was counted
# repeatedly and three quarters of the vocabulary never appeared.
#
# "Draft" is still not a status here, and sending one that does not exist is
# answered with a bare HTTP 500 that names nothing.
ISSUE_STATUSES = {
    "Open",
    "Closed",
    "Remediated",
    "Pending Remediation",
    "Inactive",
    "Canceled",
}

# The Executive Summary field, in API terms.
#
# AuditBoard exposes it as a generic custom text column; the label is applied
# by the form, not the API. Category 10's form (template 1083) places
# ``issue:customText4`` with the tooltip "Consumable summary of finding for
# stakeholders, less than 30 characters", and category 10 lists ``customText4``
# among its required-at-Open fields. Read attribute is ``custom_text4``.
#
# 30 characters is the form's guidance, not a server limit — existing values in
# the register run to 90-plus. So it is reported to the UI as a target, and the
# hard cap below is the one that actually truncates.
EXECUTIVE_SUMMARY_FIELD = "custom_text4"
EXECUTIVE_SUMMARY_TARGET = 30
MAX_EXECUTIVE_SUMMARY = 255

# Fields category 10 declares required to sit at status Open, read from
# ``issue_required_fields`` on GET /api/v1/issue_categories.
#
# The API enforces none of them: an issue created without a senior owner, a
# vice president or an identified date is accepted and stored. It just is not a
# complete register entry afterwards, which is how an issue can be filed under
# the right category and still not read as one of its records. Recorded here so
# the gap is visible in code rather than discovered in the UI.
CATEGORY_10_REQUIRED_AT_OPEN = (
    "title",
    "description",
    "customText4",
    "seniorOwnerUser",
    "vicePresidentUser",
    "identifiedDate",
)

# Issue categories that hold a standalone issue.
#
# The deciding attribute is an empty ``key``. A category with a key is bound
# to a source object — "compliance", "bcm", "regulatory-compliance-regulation-
# item" — and an issue filed under it is expected to point at one. A scanner
# finding has no AuditBoard test, control or risk behind it, so only a keyless
# category fits. Verified against GET /api/v1/issue_categories, and confirmed
# by creating issue I#1714 under 11: the server set type="standalone" and left
# issuesourceable_type null on its own.
#
# An earlier version of this list included 5, 15 and 18. That was wrong — all
# three carry a key and expect a source.
STANDALONE_CATEGORY_IDS: Dict[int, str] = {
    11: "IT Risk Exception",
    12: "IT Infrastructure Exception",
}

# Categories that are bound to a source object, and the object type each one
# expects in ``issuesourceable_type``.
#
# A keyed category is not unusable — it just needs to be told what the issue
# hangs off. Category 10 "IT Risk Issue" carries key "it-risk-mitigation-plan"
# and every one of the 427 issues already filed under it points at a Risk:
# ``issuesourceable_type: "Risk"`` with a risk ID, and the server still reports
# ``type: "standalone"`` on the result. So the key describes the register the
# category belongs to, not a requirement that the issue come from a test run.
#
# Read from the live instance rather than assumed: GET /api/v1/issue_categories
# for the keys, and the existing category-10 population for the source type.
# Only entries verified that way belong here.
SOURCE_BACKED_CATEGORIES: Dict[int, Dict[str, str]] = {
    10: {"name": "IT Risk Issue", "source_type": "Risk"},
}

MAX_TITLE = 500
MAX_DESCRIPTION = 30000


class AuditBoardError(RuntimeError):
    """Raised when AuditBoard rejects a request or is not configured."""


class AuditBoardClient:
    """Thin client over the AuditBoard issues API."""

    def __init__(self) -> None:
        self.url = (settings.AUDITBOARD_URL or "").rstrip("/")
        self.token = settings.AUDITBOARD_TOKEN or ""
        self.category_id = settings.AUDITBOARD_ISSUE_CATEGORY_ID or 0
        self.create_status = settings.AUDITBOARD_CREATE_STATUS or "Open"
        self.ca_bundle = settings.AUDITBOARD_CA_BUNDLE or ""
        self.ident_id = settings.AUDITBOARD_ISSUE_IDENT_ID or DEFAULT_ISSUE_IDENT_ID
        self.source_id = settings.AUDITBOARD_ISSUE_SOURCE_ID or 0
        self.workspace_id = settings.AUDITBOARD_WORKSPACE_ID or 0
        # AuditBoard user IDs stamped on every issue we file. Category 10
        # requires a senior owner and a vice president at status Open, and
        # neither is derivable from an AuditGitHub login — the two systems
        # share no identity. Unset means the field is omitted, which files an
        # incomplete register entry rather than failing.
        self.senior_owner_user_id = settings.AUDITBOARD_SENIOR_OWNER_USER_ID or 0
        self.vice_president_user_id = settings.AUDITBOARD_VICE_PRESIDENT_USER_ID or 0
        self.tester_user_id = settings.AUDITBOARD_TESTER_USER_ID or 0
        self.reviewer_user_id = settings.AUDITBOARD_REVIEWER_USER_ID or 0

    @property
    def source_type(self) -> Optional[str]:
        """The ``issuesourceable_type`` this category needs, if it needs one."""
        spec = SOURCE_BACKED_CATEGORIES.get(self.category_id)
        return spec["source_type"] if spec else None

    @property
    def category_name(self) -> Optional[str]:
        if self.category_id in STANDALONE_CATEGORY_IDS:
            return STANDALONE_CATEGORY_IDS[self.category_id]
        spec = SOURCE_BACKED_CATEGORIES.get(self.category_id)
        return spec["name"] if spec else None

    @property
    def enabled(self) -> bool:
        return bool(self.url and self.token and self.category_id)

    def config_problem(self) -> Optional[str]:
        """Why the client cannot be used, in words a UI can show."""
        if not self.url:
            return "AUDITBOARD_URL is not set"
        if not self.token:
            return "AUDITBOARD_TOKEN is not set"
        if not self.category_id:
            return "AUDITBOARD_ISSUE_CATEGORY_ID is not set"
        known = set(STANDALONE_CATEGORY_IDS) | set(SOURCE_BACKED_CATEGORIES)
        if self.category_id not in known:
            # Not "this category is invalid" — it is a category nobody has
            # verified a create against. Filing blind into a GRC register is
            # how you get a record you cannot delete.
            return (
                f"Issue category {self.category_id} has not been verified for "
                "filing from this app. Verified categories: "
                + ", ".join(
                    f"{k} {v}" for k, v in sorted(STANDALONE_CATEGORY_IDS.items())
                )
                + ", "
                + ", ".join(
                    f"{k} {v['name']} (needs AUDITBOARD_ISSUE_SOURCE_ID)"
                    for k, v in sorted(SOURCE_BACKED_CATEGORIES.items())
                )
            )
        if self.source_type and not self.source_id:
            spec = SOURCE_BACKED_CATEGORIES[self.category_id]
            return (
                f"Issue category {self.category_id} ({spec['name']}) attaches "
                f"each issue to a {spec['source_type']} record, and "
                "AUDITBOARD_ISSUE_SOURCE_ID is not set. Set it to the "
                f"{spec['source_type'].lower()} ID these findings belong under."
            )
        if self.create_status not in ISSUE_STATUSES:
            return (
                f"AUDITBOARD_CREATE_STATUS={self.create_status!r} is not a status "
                "on this instance. Accepted: " + ", ".join(sorted(ISSUE_STATUSES))
            )
        if not self.ca_bundle:
            return (
                "AUDITBOARD_CA_BUNDLE is not set — refusing to send a bearer "
                "token over a connection that cannot be verified"
            )
        return None

    # -- transport ---------------------------------------------------------

    def _verify(self):
        """TLS verification target.

        Egress to this host takes one of two paths depending on the moment:
        intercepted by Zscaler, which presents its own chain, or direct to
        AWS, which presents Amazon's. The "Basic Constraints of CA cert not
        marked critical" failure that sec-diligence worked around with
        ``verify=False`` is the Zscaler chain. The bundle therefore carries a
        real trust store plus the Zscaler root.

        Switching verification off is not the fix: the request carries a
        bearer token, and an unverified connection means anything in path can
        read it. Note that Zscaler terminates TLS, so the token is already
        readable by that infrastructure — verification protects it from
        everyone else, which is the part still worth having.
        """
        if not self.ca_bundle:
            raise AuditBoardError(
                "AUDITBOARD_CA_BUNDLE is not set. Refusing to transmit the "
                "AuditBoard token without verifying the server certificate."
            )
        return self.ca_bundle

    def _headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self.token}",
            "Accept": "application/json",
            "Content-Type": "application/json",
        }

    # -- payload -----------------------------------------------------------

    def build_payload(
        self,
        *,
        title: str,
        description: str,
        severity: str,
        deficiency_level_id: Optional[int] = None,
        status: Optional[str] = None,
        executive_summary: Optional[str] = None,
        identified_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Assemble the create body. Pure — safe to show a user before sending.

        Verified against the live instance on 2026-09-15 by creating and then
        closing issue I#1714. Three things the API does not document publicly
        and that a wrong guess returns a bare HTTP 500 for:

        *   ``issuesourceable_type`` is never the literal "standalone".
            ``type: "standalone"`` is something the server reports, not
            something you ask for, and sending it as a sourceable type is
            rejected with no explanation. For a keyless category (11, 12) the
            sourceable fields are omitted entirely and the server leaves them
            null. For a keyed category they name a real record — "Risk" and a
            risk ID for category 10 — which is how the 427 issues already in
            that category are shaped.
        *   ``status`` must be one of this instance's real statuses, listed in
            :data:`ISSUE_STATUSES`. "Draft" is not one of them.
        *   ``issue_ident_id`` records how the issue was identified. 7 is
            "Other", which is the honest answer for a scanner.

        ``identified_date`` can only be set here, at create. Verified on
        2026-09-16: a PUT of that one field onto an existing issue is refused
        with ``issue:action.editDraft``, while ``custom_text4`` and the four
        user stamps all accept a single-field PUT under this token. So a
        filing that goes out without a date cannot be corrected later by this
        integration — it needs someone with edit rights in the UI.

        The body is wrapped by :meth:`create_issue`, not here, so what this
        returns is exactly the field set a reviewer sees.
        """
        level = deficiency_level_id or DEFICIENCY_LEVEL_BY_SEVERITY.get(
            (severity or "").lower(), 1
        )
        if level not in DEFICIENCY_LEVELS:
            raise AuditBoardError(f"Unknown deficiency_level_id: {level}")

        status = status or self.create_status
        if status not in ISSUE_STATUSES:
            raise AuditBoardError(
                f"Unknown status {status!r}. This instance accepts: "
                + ", ".join(sorted(ISSUE_STATUSES))
            )

        payload: Dict[str, Any] = {
            "title": (title or "Untitled finding")[:MAX_TITLE],
            "description": (description or "")[:MAX_DESCRIPTION],
            "status": status,
            "issue_category_id": self.category_id,
            "issue_ident_id": self.ident_id,
            "deficiency_level_id": level,
        }
        if self.source_type and self.source_id:
            payload["issuesourceable_type"] = self.source_type
            payload["issuesourceable_id"] = self.source_id

        # Executive Summary. Required by category 10 and blank on everything
        # this app has filed so far, which is the difference between an issue
        # in the register and an issue a reader of the register can use.
        summary = (executive_summary or "").strip()
        if summary:
            payload[EXECUTIVE_SUMMARY_FIELD] = summary[:MAX_EXECUTIVE_SUMMARY]

        if identified_date:
            payload["identified_date"] = identified_date

        # Stamped only when configured. An issue with the wrong executive on
        # it is worse than one with none: it routes work to someone who never
        # agreed to own it.
        for field, value in (
            ("senior_owner_user_id", self.senior_owner_user_id),
            ("vice_president_user_id", self.vice_president_user_id),
            ("tester_user_id", self.tester_user_id),
            ("reviewer_user_id", self.reviewer_user_id),
        ):
            if value:
                payload[field] = value

        return payload

    # -- calls -------------------------------------------------------------

    def create_issue(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """POST a new issue. Returns the created issue as AuditBoard reports it."""
        problem = self.config_problem()
        if problem:
            raise AuditBoardError(problem)

        endpoint = f"{self.url}{API_PATH}"
        try:
            resp = requests.post(
                endpoint,
                # Rails-style root key. An unwrapped body is answered with a
                # bare HTTP 500 that names nothing. The API advertises the key
                # itself in every list response: meta.resourceNames.camelized.
                json={"issue": payload},
                headers=self._headers(),
                verify=self._verify(),
                timeout=30,
            )
        except requests.RequestException as exc:
            logger.error("AuditBoard request failed: %s", exc)
            raise AuditBoardError(f"AuditBoard request failed: {exc}") from exc

        if resp.status_code >= 400:
            # Body can name the rejected field, which is the whole value of
            # the message. It does not echo the token.
            detail = resp.text[:500]
            logger.warning(
                "AuditBoard rejected issue creation (%s): %s", resp.status_code, detail
            )
            raise AuditBoardError(
                f"AuditBoard returned {resp.status_code}: {detail}"
            )

        body = resp.json() if resp.content else {}
        # Create answers with the same envelope as a list: {"issues": [ ... ]}.
        issues = body.get("issues") if isinstance(body, dict) else None
        if isinstance(issues, list) and issues:
            return issues[0]
        if isinstance(body, dict):
            return body.get("issue") or body
        return {}

    def issue_url(self, issue_id: Any) -> Optional[str]:
        """Human-facing link to an issue, for reporting back to the filer.

        The path is ``/workspace/{n}/issues/issue/{id}``, confirmed from a
        working browser URL. An earlier version emitted ``/issues/{id}``,
        which looked fine because AuditBoard is a single-page app — every
        path returns HTTP 200 and the router sorts it out client-side, so a
        wrong link cannot be caught by fetching it.

        The workspace number is configuration rather than something derived.
        It is not in the API: ``/api/v1/workspaces`` lists three workspaces
        and excludes this one by type, ``filter[id]=5`` comes back empty, and
        neither the issue nor its source risk carries a workspace field.

        Returns None when the workspace is unset. A link that goes to the
        wrong place is worse than a badge with no link, because the reader
        cannot tell the difference until they are looking at someone else's
        issue.
        """
        if not self.url or issue_id in (None, "") or not self.workspace_id:
            return None
        return f"{self.url}/workspace/{self.workspace_id}/issues/issue/{issue_id}"

    def unstamped_required_fields(self) -> List[str]:
        """Category-required fields this app cannot fill from configuration.

        Not an error — filing succeeds without them, because the API enforces
        none of the category's required-field list. It is a completeness
        warning, so whoever files knows the record will need finishing by hand
        in AuditBoard rather than discovering it in a register review later.

        ``customText4`` is not listed: it comes from the filer per issue, not
        from configuration, so it is checked at filing time instead.
        """
        if self.category_id != 10:
            return []
        missing = []
        if not self.senior_owner_user_id:
            missing.append("seniorOwnerUser")
        if not self.vice_president_user_id:
            missing.append("vicePresidentUser")
        return missing

    def reference_data(self) -> Dict[str, List[Dict[str, Any]]]:
        """Vocabularies a UI needs to let a person override the mapping."""
        return {
            "issue_statuses": sorted(ISSUE_STATUSES),
            "deficiency_levels": [
                {"id": k, "name": v} for k, v in sorted(DEFICIENCY_LEVELS.items())
            ],
            "standalone_categories": [
                {"id": k, "name": v} for k, v in sorted(STANDALONE_CATEGORY_IDS.items())
            ],
            "source_backed_categories": [
                {"id": k, "name": v["name"], "source_type": v["source_type"]}
                for k, v in sorted(SOURCE_BACKED_CATEGORIES.items())
            ],
            "issue_idents": [
                {"id": k, "name": v} for k, v in sorted(ISSUE_IDENTS.items())
            ],
        }


auditboard_client = AuditBoardClient()

File: src/web-ui/lib/auditboard.ts
/**
 * AuditBoard-specific fields for filing a finding as a GRC issue.
 *
 * The issue text is built in `lib/finding-issue.ts`, shared with Jira. What is
 * AuditBoard's alone lives here: the deficiency-level vocabulary and the shape
 * of the create request.
 *
 * Two differences from the Jira path drive this module:
 *
 * 1.  There is no prefill-a-form URL. Creating an issue is an authenticated
 *     API write, so the request goes to our own API and the credential stays
 *     server-side. The browser never holds the AuditBoard token.
 * 2.  There is no confirmation screen on the far side. Pressing the button
 *     creates the record, so the dialog has to show the exact body first.
 *
 * `deficiency_level_id` is the severity-shaped field. `issue_rating_id` is
 * not — its values are control-testing outcomes (Cleared, Tested - Remains
 * Deficient, New Deficiency) and mapping a code severity onto it produces
 * nonsense. This module deliberately never sends it.
 */

import { normalizeSeverity, type Severity } from "./severity";
import {
  buildDescription,
  buildExecutiveSummary,
  buildLabels,
  buildSummary,
  occurrenceCount,
  type FindingLike,
  type IssueScope,
  type NormalizeOptions,
} from "./finding-issue";

export {
  buildExecutiveSummary,
  EXECUTIVE_SUMMARY_TARGET,
  MAX_EXECUTIVE_SUMMARY,
} from "./finding-issue";

export type {
  FindingLike,
  GroupSummary,
  IssueScope,
  OccurrenceGroup,
} from "./finding-issue";

/**
 * Tiers the AuditBoard filing endpoint accepts.
 *
 * `global` is absent on purpose. It is still a valid `IssueScope` — the
 * exception flow and the Jira deep link use it — but the filing endpoint
 * refuses it, because grouping on scanner plus file path put 1,287 unrelated
 * advisories behind one issue. Rows filed before 2026-09-16 keep that scope
 * and are still matched on it.
 */
export const FILING_TIERS = ["specific", "project", "org"] as const;
export type FilingTier = (typeof FILING_TIERS)[number];

/** Whether a scope may be sent to the filing endpoint. */
export function isFilingTier(scope: IssueScope): scope is FilingTier {
  return (FILING_TIERS as readonly string[]).includes(scope);
}

/** Label for a tier, in the words the dialog shows. */
export const FILING_TIER_LABELS: Record<FilingTier, string> = {
  specific: "This finding only",
  project: "This defect in this project",
  org: "This defect in every project",
};

/** An `{ id, name }` pair from AuditBoard's reference vocabularies. */
export interface AuditBoardRef {
  id: number;
  name: string;
}

/**
 * Response of `GET /findings/auditboard/config`.
 *
 * Carries no credential by design — the browser only learns whether the
 * integration works and which values a person may choose from.
 */
export interface AuditBoardConfig {
  enabled: boolean;
  /** Why filing is unavailable, in words a UI can show. Null when it works. */
  problem: string | null;
  base_url: string | null;
  issue_category_id: number | null;
  issue_category_name: string | null;
  /** Record type issues attach to when the category is source-backed, e.g. "Risk". */
  source_type: string | null;
  source_id: number | null;
  create_status: string;
  /** Severity → `deficiency_level_id` the server applies when none is sent. */
  default_deficiency_levels: Partial<Record<Severity, number>>;
  deficiency_levels: AuditBoardRef[];
  standalone_categories: AuditBoardRef[];
  /** Length the category's form asks the Executive Summary to stay under. */
  executive_summary_target_chars: number;
  /** True when the category will not hold a complete issue without one. */
  executive_summary_required: boolean;
  /**
   * Category-required fields the server has no configured value for — a
   * senior owner and a vice president on category 10. Filing works anyway
   * because AuditBoard enforces none of them; the record just lands
   * incomplete, so the dialog says so before it is created.
   */
  unstamped_required_fields: string[];
}

export const EMPTY_AUDITBOARD_CONFIG: AuditBoardConfig = {
  enabled: false,
  problem: null,
  base_url: null,
  issue_category_id: null,
  issue_category_name: null,
  source_type: null,
  source_id: null,
  // Not "Draft" — this instance has no such status, and a placeholder that
  // looks like a real value is worse than one that is obviously a default.
  create_status: "Open",
  default_deficiency_levels: {},
  deficiency_levels: [],
  standalone_categories: [],
  executive_summary_target_chars: 30,
  executive_summary_required: false,
  unstamped_required_fields: [],
};

/**
 * Every scope the filing endpoints can return, including the one no tier
 * describes.
 *
 * `selection` is not in {@link FILING_TIERS} on purpose: the three tiers are
 * rules, so the server can re-derive what each covers from the issue row. A
 * selection is a list of rows someone ticked, so its membership is written
 * down in `auditboard_issue_findings` instead. It is filed through its own
 * endpoint, never offered as a tier in a tier picker.
 */
export type FilingScope = FilingTier | "selection";

/** Body of `POST /findings/selection/auditboard-issue`. */
export interface AuditBoardSelectionIssueRequest {
  finding_ids: string[];
  /**
   * How many findings the filer was shown as selected. The server refuses the
   * request if a different number resolves, which is the guard against filing
   * a permanent record for a smaller set than the person saw ticked.
   */
  expected_count?: number | null;
  title: string;
  description: string;
  deficiency_level_id?: number | null;
  executive_summary?: string | null;
  /** File even when the selection holds findings below the severity floor. */
  include_ineligible?: boolean;
}

/** Body of `POST /findings/{id}/auditboard-issue`. */
export interface AuditBoardIssueRequest {
  scope: FilingTier;
  title: string;
  description: string;
  /** Omit to accept the server's severity mapping. */
  deficiency_level_id?: number | null;
  /** Written to custom_text4, the Executive Summary field. */
  executive_summary?: string | null;
}

/** Response of that POST. */
export interface AuditBoardIssueResult {
  issue_id: string | null;
  issue_url: string | null;
  deficiency_level_id: number;
  deficiency_level_name: string;
  scope: FilingScope;
  /** Counted server-side at filing time, not taken from the request. */
  occurrence_count: number;
  /** Projects the defect affected, measured at filing time. */
  project_count: number;
  /** Distinct locations written into the issue body. */
  location_count: number;
  /** Locations the body cap left out. */
  locations_omitted: number;
  /** `scanner::rule_id` the issue was filed on. Null at `specific`. */
  identity_key: string | null;
}

/** Everything the dialog previews and then sends. */
export interface AuditBoardDraft {
  title: string;
  description: string;
  /**
   * Executive Summary, in plain language. Generated as a starting point — the
   * dialog lets a person rewrite it, because the register is read by people
   * who did not file it and a phrase table cannot know the business context.
   */
  executiveSummary: string;
  severity: Severity;
  scope: FilingTier;
  /** Findings this issue speaks for, as the browser understands it. */
  occurrenceCount: number;
  /**
   * The location section the server will append, verbatim. Empty at
   * `specific`. Shown in the preview rather than folded into `description`:
   * the server writes it from the measurement it counted from, so the list
   * and the count cannot drift apart.
   */
  locationText: string;
  /** Level that will be applied — the override if set, else the default. */
  deficiencyLevelId: number | null;
  /** Name of that level, or null when the vocabulary has not loaded. */
  deficiencyLevelName: string | null;
  /** True when the filer chose a level other than the severity default. */
  deficiencyOverridden: boolean;
}

/**
 * Build the AuditBoard issue as it will be sent.
 *
 * `omitProvenance` is set because the API route appends its own footer with a
 * server-measured occurrence count and the filer's identity. Two footers that
 * can disagree about the same number is worse than one that cannot.
 *
 * Labels are not a field AuditBoard's issues API takes, so the label set goes
 * into the body as a Tags line instead — it is what makes a prior filing of
 * the same finding findable by search before someone opens a second issue.
 */
export function buildAuditBoardDraft(
  finding: FindingLike,
  options: NormalizeOptions & {
    config?: AuditBoardConfig;
    /** Filer's chosen deficiency level. Null accepts the default. */
    deficiencyLevelId?: number | null;
  } = {},
): AuditBoardDraft {
  const requested = options.scope ?? "specific";
  // A scope the filing endpoint would refuse never reaches it. 'global' rows
  // still exist and still display; they are just not written any more.
  const scope: FilingTier = isFilingTier(requested) ? requested : "specific";
  const summary = options.groupSummary ?? null;
  // Highest severity in the group, per the filing rule, so the deficiency
  // level matches what the issue actually covers rather than whichever
  // instance the filer happened to open.
  const severity =
    scope === "specific"
      ? normalizeSeverity(finding.severity)
      : normalizeSeverity(summary?.severity ?? finding.severity);
  const config = options.config ?? EMPTY_AUDITBOARD_CONFIG;

  const scopedOptions: NormalizeOptions = { ...options, scope };
  const tags = buildLabels(finding, scopedOptions);
  const normalizeOpts: NormalizeOptions = {
    ...scopedOptions,
    omitProvenance: true,
    extraSections: [
      ...(options.extraSections ?? []),
      `Tags\n${tags.join(" ")}`,
    ],
  };

  const defaultLevel = config.default_deficiency_levels?.[severity] ?? null;
  const chosen = options.deficiencyLevelId ?? null;
  const effective = chosen ?? defaultLevel;
  const name =
    config.deficiency_levels.find((l) => l.id === effective)?.name ?? null;

  return {
    title: buildSummary(finding, normalizeOpts),
    description: buildDescription(finding, normalizeOpts),
    executiveSummary: buildExecutiveSummary(finding, scopedOptions),
    severity,
    scope,
    occurrenceCount: occurrenceCount(scope, options.group, summary),
    locationText: scope === "specific" ? "" : (summary?.location_text ?? ""),
    deficiencyLevelId: effective,
    deficiencyLevelName: name,
    deficiencyOverridden: chosen !== null && chosen !== defaultLevel,
  };
}

/**
 * Levels that assert a SOX determination rather than describe a weakness.
 *
 * "Significant Deficiency" and "Material Weakness" are terms of art with
 * reporting consequences. A static-analysis hit is evidence of a control
 * weakness, not a determination of one, so the server never maps a severity
 * onto these — a person has to choose them, and the dialog says why.
 */
export const ESCALATED_DEFICIENCY_LEVEL_IDS = new Set([3, 4]);

// ---------------------------------------------------------------------------
// Filing index — annotating a list of findings with the issue each one became
// ---------------------------------------------------------------------------

/** One row of `GET /findings/auditboard/filings`. */
export interface AuditBoardFiling {
  finding_id: string;
  /** Public finding ID. Null once the finding itself has been deleted. */
  finding_uuid: string | null;
  scanner_name: string | null;
  file_path: string | null;
  /** Rule the defect is keyed on. Null on pre-2026-09-16 `global` rows. */
  rule_id: string | null;
  /** Project a `project` row is confined to. Null at every other tier. */
  repository_id: string | null;
  /**
   * Every `scanner::rule_id` this row covers, expanded server-side for
   * reviewer-approved cross-scanner merges. Matching against this rather than
   * against `rule_id` is what keeps the browser out of the equivalence table.
   */
  identity_keys: string[];
  scope: IssueScope;
  issue_id: string;
  issue_uid: string | null;
  issue_url: string | null;
  issue_status: string | null;
  deficiency_level_name: string | null;
  occurrence_count: number;
  /** Projects the defect affected when filed. */
  project_count: number | null;
  filed_by: string | null;
  filed_at: string | null;
}

/** A filing matched to a finding, and how it matched. */
export interface FilingMatch {
  filing: AuditBoardFiling;
  /** True when this finding is the one the issue was filed from. */
  direct: boolean;
}

export interface FilingIndex {
  byFinding: Map<string, AuditBoardFiling>;
  /** `project` rows, keyed by identity and project. */
  byProject: Map<string, AuditBoardFiling>;
  /** `org` rows, keyed by identity alone. */
  byIdentity: Map<string, AuditBoardFiling>;
  /** Pre-2026-09-16 `global` rows, keyed by scanner and path. */
  byLegacyGroup: Map<string, AuditBoardFiling>;
}

/** Defect identity. Must stay identical to `Identity.key` on the server. */
export function filingIdentityKey(
  scannerName: string | null | undefined,
  ruleId: string | null | undefined,
): string {
  return `${scannerName ?? ""}::${ruleId ?? ""}`;
}

/** Identity plus project, for a `project`-tier row. */
export function filingProjectKey(
  identityKey: string,
  repositoryId: string | null | undefined,
): string {
  return `${identityKey}@${repositoryId ?? ""}`;
}

/**
 * Legacy group key: scanner + path.
 *
 * Kept because rows filed before 2026-09-16 carry it, and nothing in
 * AuditBoard can be deleted — those issues have to keep matching their
 * findings. Never written to a new row.
 */
export function filingGroupKey(
  scannerName: string | null | undefined,
  filePath: string | null | undefined,
): string {
  return `${scannerName ?? ""}\u0000${filePath ?? ""}`;
}

/**
 * Index filings for per-row lookup.
 *
 * Built once for a whole table rather than queried per row: a findings list
 * runs to thousands of rows while the filing table holds one row per issue,
 * so one fetch and a few maps beat thousands of requests.
 *
 * A row is indexed under every identity it covers, not only its own, so a
 * reviewer-approved cross-scanner merge annotates the other scanner's
 * findings too.
 */
export function buildFilingIndex(rows: AuditBoardFiling[]): FilingIndex {
  const byFinding = new Map<string, AuditBoardFiling>();
  const byProject = new Map<string, AuditBoardFiling>();
  const byIdentity = new Map<string, AuditBoardFiling>();
  const byLegacyGroup = new Map<string, AuditBoardFiling>();

  for (const row of rows) {
    if (row.finding_uuid) byFinding.set(row.finding_uuid, row);

    if (row.scope === "global") {
      byLegacyGroup.set(filingGroupKey(row.scanner_name, row.file_path), row);
      continue;
    }

    const keys =
      row.identity_keys?.length > 0
        ? row.identity_keys
        : [filingIdentityKey(row.scanner_name, row.rule_id)];

    if (row.scope === "project") {
      for (const key of keys) byProject.set(filingProjectKey(key, row.repository_id), row);
    } else if (row.scope === "org") {
      for (const key of keys) byIdentity.set(key, row);
    }
  }

  return { byFinding, byProject, byIdentity, byLegacyGroup };
}

/**
 * Find the issue covering a finding, if any.
 *
 * A direct match wins over a group match when both exist: "this was filed"
 * and "something else covers this" are different statements, and the stronger
 * one is the one worth showing. Group matches are tried narrowest first —
 * project, then organization, then the retired scanner-and-path key — so the
 * issue shown is the one whose scope sits closest to the finding in hand.
 */
export function lookupFiling(
  index: FilingIndex | null,
  finding: {
    id?: string;
    scanner_name?: string | null;
    rule_id?: string | null;
    repository_id?: string | null;
    file_path?: string | null;
  },
): FilingMatch | null {
  if (!index) return null;

  const direct = finding.id ? index.byFinding.get(finding.id) : undefined;
  if (direct) return { filing: direct, direct: true };

  if (finding.rule_id) {
    const identity = filingIdentityKey(finding.scanner_name, finding.rule_id);
    const project = index.byProject.get(filingProjectKey(identity, finding.repository_id));
    if (project) return { filing: project, direct: false };
    const org = index.byIdentity.get(identity);
    if (org) return { filing: org, direct: false };
  }

  const legacy = index.byLegacyGroup.get(
    filingGroupKey(finding.scanner_name, finding.file_path),
  );
  return legacy ? { filing: legacy, direct: false } : null;
}
